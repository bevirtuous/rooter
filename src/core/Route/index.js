import { parseUrl } from 'query-string';

const parseOptions = {
  arrayFormat: 'comma',
  parseBooleans: true,
  parseNumbers: true,
};

function Route(options) {
  const { meta = {} } = options;
  const location = options.pathname.trim();
  const id = options.id;

  const splitPath = location.split('#');
  const path = splitPath[0];
  const hash = splitPath[1] || null;
  const { query, url: pathname } = parseUrl(path, parseOptions);

  return {
    id,
    hash,
    location,
    meta,
    pathname,
    query,
  };
}

export default Route;
