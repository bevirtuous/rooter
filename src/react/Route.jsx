import React from 'react';
import Context from './context';
import usePath from './usePath';
import useLocation from './useLocation';

function Route({ children, exact = false, path }) {
  const location = useLocation();
  const { isExactMatch, isMatch, params } = usePath(path);
  const isVisible = exact ? isExactMatch : isMatch;

  if (!isVisible) {
    return null;
  }

  return (
    <Context.Provider value={{ ...location, params, pattern: path }}>
      {children}
    </Context.Provider>
  );
}

export default Route;