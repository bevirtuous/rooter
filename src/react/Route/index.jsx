import React from 'react';
import PropTypes from 'prop-types';
import router from '../../core/Router';
import RouteNotFound from '../404';
import { RouteContext, RouterContext } from '../context';

class Route extends React.Component {
  constructor(props) {
    super(props);
    router.register(props.path);
  }

  render() {
    const { children, path } = this.props;
    const route = router.getCurrentRoute();

    if (route.pattern !== path) {
      return null;
    }

    const key = `${route.id}-${route.pathname}`;

    return (
      <RouteContext.Provider key={key} value={route}>
        {children}
      </RouteContext.Provider>
    );
  }
}

Route.contextType = RouterContext;
Route.NotFound = RouteNotFound;
Route.propTypes = {
  children: PropTypes.node.isRequired,
  path: PropTypes.string.isRequired,
};

export default Route;
