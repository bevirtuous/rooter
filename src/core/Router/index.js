import Route from '../Route';
import * as errors from './errors';
import emitter from '../emitter';
import history from '../history';
import matcher from '../matcher';
import stack from '../Stack';
import * as constants from '../constants';

class Router {
  constructor(createHistory = history) {
    // Flag to indicate a native history event. Should always be reset to true.
    this.nativeEvent = true;

    this.history = createHistory();

    // The patterns are collected to match against pathnames.
    this.patterns = {};

    // The `currentIndex` is used to track which stack entry is the current route.
    this.currentIndex = 0;

    this.action = constants.PUSH;

    // Unsubscribe to any other history module changes.
    if (typeof this.historyListener === 'function') {
      this.historyListener();
      stack.clear();
    }

    this.addInitialRoute();

    this.historyListener = this.history.listen(this.handleNativeEvent);
  }

  handleNativeEvent = (location, action) => {
    if (!this.nativeEvent) {
      return;
    }

    const next = stack.getByIndex(this.currentIndex + 1);

    if (next && location.state && location.state.route && next.id === location.state.route.id) {
      this.handlePush({
        to: location.pathname,
        meta: location.state,
      }, false, true);

      return;
    }

    if (action === constants.POP) {
      this.handlePop();
    }
  }

  /**
   * Populate the stack with an initial entry to match the history module.
   * Note: we cannot match it against a pattern at this point.
   */
  addInitialRoute = () => {
    const { hash, pathname, search } = this.history.location;
    const fullPathname = `${pathname}${search}${hash}`;
    const route = new Route({ pathname: fullPathname });

    stack.add(route.id, route);
  }

  /**
   * @param {Object} params The router action params.
   * @returns {Promise}
   */
  handlePop = (params = {}) => new Promise((resolve, reject) => {
    const {
      emit = true,
      forceNative = false,
      steps = 1,
      meta = null,
    } = params;
    let unlisten = null;

    if (steps <= 0) {
      reject(new Error(errors.EINVALIDSTEPS));
      this.nativeEvent = true;
      return;
    }

    // Get id of target route.
    const targetIndex = Math.max(this.currentIndex - steps, 0);
    const prev = stack.getByIndex(this.currentIndex);
    const next = stack.getByIndex(targetIndex);

    if (meta) {
      next.meta = Object.assign(next.meta, meta);
    }

    const end = { action: constants.POP, prev, next };

    /**
     *
     */
    const callback = () => {
      unlisten();
      this.currentIndex = targetIndex;
      this.action = constants.POP;

      if (emit) {
        emitter.emit(constants.EVENT, end);
      }

      resolve(end);
      this.nativeEvent = true;
    };

    /**
     * Create a reference to the history listener
     * to be able to unsubscribe from inside the callback.
     */
    unlisten = this.history.listen(callback);

    // Perform the history back action.
    if (forceNative || !this.nativeEvent) {
      this.history.go(steps * -1);
    } else {
      callback();
    }
  });

  /**
   * @param {Object} params The params to use when navigating.
   * @param {boolean} [cleanStack=true] When true, the overhanging routes will be removed.
   * @param {boolean} [nativePush=false] When true, the process of adding a route to
   * the stack is skipped.
   * @returns {Promise}
   */
  handlePush(params, cleanStack = true, nativePush = false) {
    return new Promise((resolve, reject) => {
      // Check for missing parameters.
      if (!params) {
        reject(new Error(errors.EPARAMSMISSING));
        this.nativeEvent = true;
        return;
      }

      // Check for empty params.
      if (Object.keys(params).length === 0) {
        reject(new Error(errors.EPARAMSEMPTY));
        this.nativeEvent = true;
        return;
      }

      const {
        emit = true,
        meta,
        to,
      } = params;
      const pattern = this.findPattern(to.split('?')[0]);
      let unlisten = null;

      // Remove all unwanted items from the stack.
      if (cleanStack) {
        while (this.currentIndex < stack.getAll().size - 1) {
          const [id] = stack.last();
          stack.remove(id);
        }
      }

      const prev = stack.getByIndex(this.currentIndex);
      const next = nativePush ? stack.getByIndex(this.currentIndex + 1) : new Route({
        pathname: to,
        pattern,
        meta,
      });

      if (!nativePush) {
        // Add item to the stack
        stack.add(next.id, next);
      }

      /**
       * The history event callback.
       */
      const callback = () => {
        // Unsubscribe from the history events.
        unlisten();

        // Increment the route index.
        this.currentIndex += 1;

        this.action = constants.PUSH;

        // Emit completion event.
        if (emit) {
          emitter.emit(constants.EVENT, { action: constants.PUSH, prev, next });
        }

        // Resolve the Promise.
        resolve({ prev, next });
        this.nativeEvent = true;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history push action.
      if (!this.nativeEvent) {
        this.history.push({
          pathname: to,
          state: {
            ...meta,
            route: { id: next.id },
          },
        });
      } else {
        callback();
      }
    });
  }

  /**
   * Match the given pathname to a registered pattern.
   * @param {string} pathname The pathname to match.
   * @returns {string|null}
   */
  findPattern = (pathname) => {
    const pattern = Object.keys(this.patterns).find((key) => this.patterns[key].match(pathname));
    return pattern || null;
  }

  /**
   * Registers a route pattern to match new pathnames against.
   * @param {string} pattern The pattern to register.
   */
  register = (pattern) => {
    if (!pattern) {
      throw new Error(errors.EMISSINGPATTERN);
    }

    if (typeof pattern !== 'string') {
      throw new Error(errors.EINVALIDPATTERN);
    }

    const match = matcher(pattern);

    this.patterns[pattern] = {
      match,
    };

    // Find the pathname of the first route.
    const route = stack.first()[1];

    // If it has been set then we don't need to match it.
    if (route.pattern !== null) {
      return;
    }

    //
    if (match(route.pathname)) {
      route.setPattern(pattern);

      const next = { action: constants.PUSH, prev: null, next: route };

      emitter.emit(constants.EVENT, next, true);
    }
  }

  deregister = (pattern) => {
    delete this.patterns[pattern];
  }

  handleReplace = (params) => new Promise((resolve, reject) => {
    // Check for missing parameters.
    if (!params) {
      reject(new Error(errors.EPARAMSMISSING));
      this.nativeEvent = true;
      return;
    }

    // Check for empty params.
    if (Object.keys(params).length === 0) {
      reject(new Error(errors.EPARAMSEMPTY));
      this.nativeEvent = true;
      return;
    }

    const {
      emit = true,
      to,
      meta,
    } = params;
    const pattern = this.findPattern(to.split('?')[0]);
    let unlisten = null;

    const { id } = stack.getByIndex(this.currentIndex);
    const prev = stack.get(id);
    const next = new Route({
      pathname: to,
      pattern,
      meta,
    });
    const end = { action: constants.REPLACE, prev, next };

    // Add item to the stack.
    stack.add(id, next);

    /**
     * The history event callback.
     */
    const callback = () => {
      // Unsubscribe from the history events.
      unlisten();

      this.action = constants.REPLACE;

      // Emit completion event.
      if (emit) {
        emitter.emit(constants.EVENT, end);
      }

      resolve(end);
      this.nativeEvent = true;
    };

    /**
     * Create a reference to the history listener
     * to be able to unsubscribe from inside the callback.
     */
    unlisten = this.history.listen(callback);

    // Perform the history replace action.
    if (!this.nativeEvent) {
      this.history.replace({
        pathname: to,
        state: {
          ...meta,
          route: { id },
        },
      });
    } else {
      callback();
    }
  })

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  push = (params) => {
    this.nativeEvent = false;
    return this.handlePush(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  pop = (params) => {
    this.nativeEvent = false;
    return this.handlePop(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  replace = (params) => {
    this.nativeEvent = false;
    return this.handleReplace(params);
  }

  /**
   * @returns {Promise}
   */
  reset = ({ to = null, meta = null } = {}) => new Promise((resolve, reject) => {
    if (this.currentIndex === 0) {
      reject();
      return;
    }

    this.nativeEvent = false;
    const route = stack.first()[1];

    const prev = stack.getByIndex(this.currentIndex);
    const next = {
      action: constants.RESET,
      prev,
      next: route,
    };

    if (meta) {
      this.update(route.id, meta, false);
    }

    const params = {
      emit: false,
      forceNative: true,
      steps: this.currentIndex,
    };

    this.handlePop(params)
      .then(() => {
        if (!to) {
          emitter.emit(constants.EVENT, next);
          this.nativeEvent = true;
          resolve(next);
          return;
        }

        this.handleReplace({ to, meta, emit: false }).then(() => {
          next.next = stack.first()[1];

          this.nativeEvent = true;
          emitter.emit(constants.EVENT, next);
          resolve(next);
        });
      });
  });

  /**
   * @param {string} id The route id to update.
   * @param {Object} meta The new meta.
   * @param {boolean} emit When true, will emit when the meta was updated.
   * @returns {Promise}
   */
  update = (id, meta = {}, emit = true) => new Promise((resolve, reject) => {
    if (!id || Object.keys(meta).length === 0) {
      reject(new Error(errors.EPARAMSINVALID));
      return;
    }

    const route = stack.get(id);

    if (!route) {
      reject(new Error(errors.EINVALIDID));
      return;
    }

    route.meta = Object.assign(route.meta, meta);
    route.updated = Date.now();

    stack.update(id, route);

    if (emit) {
      emitter.emit(constants.EVENT, { action: constants.UPDATE, route });
    }

    resolve(route);
  });

  /**
   * @returns {Route}
   */
  getCurrentRoute = () => stack.getByIndex(this.currentIndex)

  /**
   * Returns the matches pattern for the given pathname.
   * @param {string} pathname The pathname to match.
   * @returns {string|null}
   */
  match = (pathname = null) => {
    if (!pathname) {
      return false;
    }

    let foundPattern = false;

    Object.entries(this.patterns).some(([pattern, properties]) => {
      if (properties.match(pathname)) {
        foundPattern = pattern;
        return true;
      }

      return false;
    });

    return foundPattern;
  }
}

export default new Router();
