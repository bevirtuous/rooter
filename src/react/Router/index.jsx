import React, { useEffect, useState } from 'react';
import router from '../../core/Router';
import { RouterContext } from '../context';
import { EVENT, UPDATE } from '../../core/constants';
import emitter from '../../core/emitter';

function Router({ children }) {
  const [routes, setRoutes] = useState({
    prev: null,
    next: router.getCurrentRoute(),
  });

  function handleChange({ prev, next }) {
    setRoutes({
      prev,
      next,
    });
  }

  function handleUpdate(next) {
    setRoutes((old) => ({
      prev: old.prev,
      next,
    }));
  }

  useEffect(() => {
    function handler(params) {
      if (params.action === UPDATE) {
        handleUpdate(params.route);
      } else {
        handleChange(params);
      }
    }

    emitter.on(EVENT, handler);

    return () => emitter.off(EVENT, handler);
  }, []);

  return (
    <RouterContext.Provider value={routes}>
      {children}
    </RouterContext.Provider>
  );
}

Router.defaultProps = {
  history: null,
};

export default Router;
