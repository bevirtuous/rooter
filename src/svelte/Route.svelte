<script>
  import { register } from '../core';
  import store from './store';

  export let component;
  export let path;

  register(path);
</script>

{#if $store.current && path === $store.current.pattern}
  <svelte:component this={component} />
{/if}
