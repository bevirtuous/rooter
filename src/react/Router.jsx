import { useEffect } from 'react';
import history from 'history/browser';
import useStore from './store';

function Router({ children }) {
  const back = useStore((state) => state.back);
  const push = useStore((state) => state.push);
  const nextRoute = useStore((state) => state.routes[state.current + 1]);

  useEffect(() => {
    const listener = history.listen(({ action, location }) => {
      if (action === 'PUSH' || (nextRoute && location.key === nextRoute.id)) {
        push(location);
      } else {
        back();
      }
    });

    return () => listener();
  }, [back, nextRoute, push]);

  return children;
}

export default Router;
