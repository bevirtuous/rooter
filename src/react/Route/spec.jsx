import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import router from '../../core/Router';
import Router from '../Router/index';
import { RouteContext } from '../context';
import Route from './index';

const path = '/myroute/:id';
const spy = jest.spyOn(router, 'register');

describe('<Route />', () => {
  it('should render component', () => {
    const app = mount((
      <Router>
        <Route path={path}>
          <div />
        </Route>
      </Router>
    ));

    expect(spy).toHaveBeenCalledWith(path);
    expect(app).toMatchSnapshot();
  });

  it('should correctly set the RouteContext value', () => {
    let contextValue = null;
    const current = router.getCurrentRoute();

    const MyComponent = () => (
      <RouteContext.Consumer>
        {(route) => {
          contextValue = route;
          return null;
        }}
      </RouteContext.Consumer>
    );

    mount((
      <Router>
        <Route path={path}>
          <MyComponent />
        </Route>
      </Router>
    ));

    expect(current).toMatchObject(contextValue);
  });

  it('should render null when does not match current route', () => {
    const route = mount((
      <Router>
        <Route path="/wrong">
          <div />
        </Route>
      </Router>
    ));

    expect(route.html()).toBe('');
  });

  it('should react to router events and update', async () => {
    const route = mount((
      <Router>
        <Route path={path}>
          <div />
        </Route>
        <Route path="/other">
          <div />
        </Route>
      </Router>
    ));

    // Should initially render the component (match found).
    expect(route.find(Route).at(0).html()).toBe('<div></div>');
    expect(route.find(Route).at(1).html()).toBeNull();

    await act(async () => {
      await router.push({ to: '/other' });
    });

    route.update();

    // Should now render null (first route does not match).
    expect(route.find(Route).at(0).html()).toBeNull();
    expect(route.find(Route).at(1).html()).toBe('<div></div>');
  });
});
