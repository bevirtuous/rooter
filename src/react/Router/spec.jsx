import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import emitter from '../../core/emitter';
import router from '../../core/Router';
import { EVENT } from '../../core/constants';
import Router from './index';
import useHistory from '../useHistory';

describe('<Router />', () => {
  it('should render with children and context', () => {
    const app = mount((
      <Router>
        <p />
      </Router>
    ));

    expect(app).toMatchSnapshot();
  });

  it('should unsubscribe when unmounted', () => {
    emitter.removeAllListeners([EVENT]);

    const app = mount((
      <Router>
        <p />
      </Router>
    ));

    expect(emitter.listenerCount(EVENT)).toBe(1);
    app.unmount();
    expect(emitter.listenerCount(EVENT)).toBe(0);
  });

  it('should update on navigation action', async () => {
    const MyComponent = () => useHistory().current.pathname;

    const app = mount((
      <Router>
        <MyComponent />
      </Router>
    ));

    expect(app.html()).toBe('/myroute/123');

    await act(async () => {
      await router.push({ to: '/myroute/456' });
    });

    app.update();

    expect(app.html()).toBe('/myroute/456');
  });

  it('should update on route update', async () => {
    const MyComponent = () => useHistory().current.meta.test;

    await act(async () => {
      await router.push({ to: '/myroute/456', meta: { test: 123 } });
    });

    const app = mount((
      <Router>
        <MyComponent />
      </Router>
    ));

    const { id } = router.getCurrentRoute();

    expect(app.html()).toBe('123');

    await act(async () => {
      await router.update(id, { test: 789 });
    });

    app.update();

    expect(app.html()).toBe('789');
  });
});
