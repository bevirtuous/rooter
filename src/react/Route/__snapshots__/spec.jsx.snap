// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Route /> should render component 1`] = `
<Router
  history={null}
>
  <Route
    path="/myroute/:id"
  >
    <div />
  </Route>
</Router>
`;
